--- CImg.h.orig	2016-05-11 16:52:09.000000000 +0200
+++ CImg.h	2016-05-11 17:08:14.000000000 +0200
@@ -23195,25 +23195,30 @@
             T *ptrd = *expression=='<'?_data + _width*_height*_depth - 1:_data;
             if (*expression=='<') {
               CImg<doubleT> res(1,mp.result_dim);
-              cimg_rofXYZ(*this,x,y,z) {
+              cimg_rofZ(*this,z) cimg_rofY(*this,y) { cimg_test_abort(); cimg_rofX(*this,x) {
                 mp(x,y,z,0,res._data);
                 const double *ptrs = res._data;
                 T *_ptrd = ptrd--; for (unsigned int n = N; n>0; --n) { *_ptrd = (T)(*ptrs++); _ptrd+=whd; }
+                }
               }
             } else if (*expression=='>' || !do_in_parallel) {
               CImg<doubleT> res(1,mp.result_dim);
-              cimg_forXYZ(*this,x,y,z) {
+              cimg_forZ(*this,z) cimg_forY(*this,y) { cimg_test_abort(); cimg_forX(*this,x) {
                 mp(x,y,z,0,res._data);
                 const double *ptrs = res._data;
                 T *_ptrd = ptrd++; for (unsigned int n = N; n>0; --n) { *_ptrd = (T)(*ptrs++); _ptrd+=whd; }
+                }
               }
             } else {
 #ifdef cimg_use_openmp
+              bool go = true;
 #pragma omp parallel
               {
                 _cimg_math_parser _mp = omp_get_thread_num()?mp:_cimg_math_parser(), &lmp = omp_get_thread_num()?_mp:mp;
 #pragma omp for collapse(2)
                 cimg_forYZ(*this,y,z) {
+                  if (go) { try {
+                  cimg_test_abort();
                   CImg<doubleT> res(1,lmp.result_dim);
                   T *ptrd = data(0,y,z,0);
                   cimg_forX(*this,x) {
@@ -23221,6 +23226,10 @@
                     const double *ptrs = res._data;
                     T *_ptrd = ptrd++; for (unsigned int n = N; n>0; --n) { *_ptrd = (T)(*ptrs++); _ptrd+=whd; }
                   }
+                  } catch (...) {
+#pragma omp atomic
+                    go &= false;
+                  } }
                 }
               }
 #endif
@@ -23228,23 +23237,31 @@
 
           } else { // Scalar-valued expression
             T *ptrd = *expression=='<'?end() - 1:_data;
-            if (*expression=='<')
-              cimg_rofXYZC(*this,x,y,z,c) *(ptrd--) = (T)mp(x,y,z,c);
-            else if (*expression=='>' || !do_in_parallel)
-              cimg_forXYZC(*this,x,y,z,c) *(ptrd++) = (T)mp(x,y,z,c);
-            else {
+            if (*expression=='<') {
+              cimg_rofC(*this,c) cimg_rofZ(*this,z) cimg_rofY(*this,y) { cimg_test_abort(); cimg_rofX(*this,x)  *(ptrd--) = (T)mp(x,y,z,c); }
+            } else if (*expression=='>' || !do_in_parallel) {
+              cimg_forC(*this,c) cimg_forZ(*this,z) cimg_forY(*this,y) { cimg_test_abort(); cimg_forX(*this,x) *(ptrd++) = (T)mp(x,y,z,c); }
+            } else {
 #ifdef cimg_use_openmp
+              bool go = true;
 #pragma omp parallel
               {
                 _cimg_math_parser _mp = omp_get_thread_num()?mp:_cimg_math_parser(), &lmp = omp_get_thread_num()?_mp:mp;
 #pragma omp for collapse(3)
                 cimg_forYZC(*this,y,z,c) {
+                  if (go) { try {
+                  cimg_test_abort();
                   T *ptrd = data(0,y,z,c);
                   cimg_forX(*this,x) *ptrd++ = (T)lmp(x,y,z,c);
+                  } catch (...) {
+#pragma omp atomic
+                    go &= false;
+                  } }
                 }
               }
 #endif
             }
+            cimg_test_abort();
           }
         } catch (CImgException& e) { CImg<charT>::string(e._message).move_to(is_error); }
 
@@ -29506,10 +29523,12 @@
           mx2 = mask.width()/2, my2 = mask.height()/2, mz2 = mask.depth()/2,
           mx1 = mx2 - 1 + (mask.width()%2), my1 = my2 - 1 + (mask.height()%2), mz1 = mz2 - 1 + (mask.depth()%2),
           mxe = width() - mx2, mye = height() - my2, mze = depth() - mz2;
+        bool go = true;
 #ifdef cimg_use_openmp
 #pragma omp parallel for cimg_openmp_if(res._spectrum>=2)
 #endif
         cimg_forC(res,c) {
+          if (go) { try {
           cimg_test_abort();
           const CImg<T> _img = get_shared_channel(c%_spectrum);
           const CImg<t> _mask = mask.get_shared_channel(c%mask._spectrum);
@@ -29623,6 +29642,12 @@
                 }
               }
           }
+          } catch (...) {
+#ifdef cimg_use_opemp
+#pragma omp atomic
+#endif
+            go &= false;
+          } }
         }
       }
       return res;
@@ -29753,10 +29778,12 @@
         mx2 = mask.width()/2, my2 = mask.height()/2, mz2 = mask.depth()/2,
         mx1 = mx2 - 1 + (mask.width()%2), my1 = my2 - 1 + (mask.height()%2), mz1 = mz2 - 1 + (mask.depth()%2),
         mxe = width() - mx2, mye = height() - my2, mze = depth() - mz2;
+      bool go = true;
 #ifdef cimg_use_openmp
 #pragma omp parallel for cimg_openmp_if(_spectrum>=2)
 #endif
       cimg_forC(*this,c) {
+        if (go) { try {
         cimg_test_abort();
         const CImg<T> _img = get_shared_channel(c%_spectrum);
         const CImg<t> _mask = mask.get_shared_channel(c%mask._spectrum);
@@ -29869,7 +29896,12 @@
               }
             }
         }
-      }
+      } catch (...) {
+#ifdef cimg_use_opemp
+#pragma omp atomic
+#endif
+        go &= false;
+      } } }
       return res;
     }
 
@@ -30055,10 +30087,12 @@
         mx2 = mask.width()/2, my2 = mask.height()/2, mz2 = mask.depth()/2,
         mx1 = mx2 - 1 + (mask.width()%2), my1 = my2 - 1 + (mask.height()%2), mz1 = mz2 - 1 + (mask.depth()%2),
         mxe = width() - mx2, mye = height() - my2, mze = depth() - mz2;
+      bool go = true;
 #ifdef cimg_use_openmp
 #pragma omp parallel for cimg_openmp_if(_spectrum>=2)
 #endif
       cimg_forC(*this,c) {
+        if (go) { try {
         cimg_test_abort();
         const CImg<T> _img = get_shared_channel(c%_spectrum);
         const CImg<t> _mask = mask.get_shared_channel(c%mask._spectrum);
@@ -30170,7 +30204,12 @@
               }
             }
         }
-      }
+      } catch (...) {
+#ifdef cimg_use_opemp
+#pragma omp atomic
+#endif
+        go &= false;
+      } } }
       return res;
     }
 
@@ -30182,6 +30221,7 @@
     **/
     CImg<T>& dilate(const unsigned int sx, const unsigned int sy, const unsigned int sz=1) {
       if (is_empty() || (sx==1 && sy==1 && sz==1)) return *this;
+      bool go = true;
       if (sx>1 && _width>1) { // Along X-axis.
         const int L = width(), off = 1, s = (int)sx, _s2 = s/2 + 1, _s1 = s - _s2, s1 = _s1>L?L:_s1, s2 = _s2>L?L:_s2;
         CImg<T> buf(L);
@@ -30189,6 +30229,7 @@
 #pragma omp parallel for collapse(3) firstprivate(buf) if (size()>524288)
 #endif
         cimg_forYZC(*this,y,z,c) {
+          if (go) { try {
           T *const ptrdb = buf._data, *ptrd = ptrdb, *const ptrde = buf._data + L - 1;
           const T *const ptrsb = data(0,y,z,c), *ptrs = ptrsb, *const ptrse = ptrs + L*off - off;
           T cur = *ptrs; ptrs+=off; bool is_first = true;
@@ -30222,6 +30263,12 @@
             }
             T *pd = data(0,y,z,c); cimg_for(buf,ps,T) { *pd = *ps; pd+=off; }
           }
+          } catch (...) {
+#ifdef cimg_use_opemp
+#pragma omp atomic
+#endif
+            go &= false;
+          } }
         }
       }
 
@@ -30233,6 +30280,7 @@
 #pragma omp parallel for collapse(3) firstprivate(buf) if (size()>524288)
 #endif
         cimg_forXZC(*this,x,z,c) {
+          if (go) { try {
           T *const ptrdb = buf._data, *ptrd = ptrdb, *const ptrde = buf._data + L - 1;
           const T *const ptrsb = data(x,0,z,c), *ptrs = ptrsb, *const ptrse = ptrs + L*off - off;
           T cur = *ptrs; ptrs+=off; bool is_first = true;
@@ -30266,6 +30314,12 @@
             }
             T *pd = data(x,0,z,c); cimg_for(buf,ps,T) { *pd = *ps; pd+=off; }
           }
+          } catch (...) {
+#ifdef cimg_use_opemp
+#pragma omp atomic
+#endif
+            go &= false;
+          } }
         }
       }
 
@@ -30277,6 +30331,7 @@
 #pragma omp parallel for collapse(3) firstprivate(buf) if (size()>524288)
 #endif
         cimg_forXYC(*this,x,y,c) {
+          if (go) { try {
           T *const ptrdb = buf._data, *ptrd = ptrdb, *const ptrde = buf._data + L - 1;
           const T *const ptrsb = data(x,y,0,c), *ptrs = ptrsb, *const ptrse = ptrs + L*off - off;
           T cur = *ptrs; ptrs+=off; bool is_first = true;
@@ -30310,6 +30365,12 @@
             }
             T *pd = data(x,y,0,c); cimg_for(buf,ps,T) { *pd = *ps; pd+=off; }
           }
+          } catch (...) {
+#ifdef cimg_use_opemp
+#pragma omp atomic
+#endif
+            go &= false;
+          } }
         }
       }
       return *this;
