--- CImg.h.orig	2017-06-28 15:33:13.000000000 +0200
+++ CImg.h	2017-06-28 15:34:01.000000000 +0200
@@ -34554,7 +34554,7 @@
         CImg<Tfloat> res(_width,_height,_depth,_spectrum,0), W(_width,_height,_depth,is_3d?4:3), val(_spectrum,1,1,1,0);
         int N = 0;
         if (is_3d) { // 3d version
-          for (float phi = (180%(int)da)/2.0f; phi<=180; phi+=da) {
+          for (float phi = (180-da*(int)(180/da))/2.0f; phi<=180; phi+=da) {
             const float phir = (float)(phi*cimg::PI/180), datmp = (float)(da/std::cos(phir)),
               da2 = datmp<1?360.0f:datmp;
             for (float theta = 0; theta<360; (theta+=da2),++N) {
@@ -34660,7 +34660,7 @@
             }
           }
         } else { // 2d LIC algorithm
-          for (float theta = (360%(int)da)/2.0f; theta<360; (theta+=da),++N) {
+          for (float theta = (360-da*(int)(360/da))/2.0f; theta<360; (theta+=da),++N) {
             const float thetar = (float)(theta*cimg::PI/180),
               vx = (float)(std::cos(thetar)), vy = (float)(std::sin(thetar));
             const t *pa = G.data(0,0,0,0), *pb = G.data(0,0,0,1), *pc = G.data(0,0,0,2);
